std/stdio.d

```diff
@@ -5458,17 +5456,22 @@ private struct ReadlnAppender
 }
 
 // Private implementation of readln
-private size_t readlnImpl(FILE* fps, ref char[] buf, dchar terminator, File.Orientation orientation)
+private size_t readlnImpl(FILE* fps, ref char[] buf, dchar terminator, File.Orientation orientation) @safe
 {
+    alias trusted_FLOCK = (fps) @trusted => _FLOCK(fps);
```

Comment:
This is a controversial pattern, because even though it's inside a function, it doesn't have a `@safe` interface and can cause memory corruption when called wrongly inside this `@safe` function body. (@aG0aep6G)

Without refactoring, the correct thing to do is mark the function `@trusted`. The `@trusted` lambda's may give the illusion that the surface of `@trusted` code is smaller, but you could still escape `FILE* fps` or add a wrong FLOCK/FUNLOCK call anywhere in the `@safe` body.

Reply:
> This is a controversial pattern, because even though it's inside a function, it doesn't have a `@safe` interface and can cause memory corruption when called wrongly inside this `@safe` function body. (@aG0aep6G)

I am pleased to see that my constant nagging is having an effect :)

Reply:
One way to fix this is to encapsulate all the necessary trusted functionality into a static struct that is non-copyable, and use that instead.

Reply:
Ideally that struct's data would be `@system`, but that's not in the language yet.

Reply:
@schveiguy It works with -preview=systemVariables. I have implemented this. There is still one hole though, the user can call `destroy` on the struct instance. That will unlock even though an unshared file pointer may still be active (though access is protected with `scope`, the struct is still in scope).

Reply:
The destructor should null the pointer.

Reply:
Why? `destroy` (or `move`) already does that.

Reply:
OK, I misunderstood. I didn't realize you were exposing the pointer elsewhere.